gateway-sp-comms/src/single_sp.rs

```diff
@@ -0,0 +1,749 @@
+            tokio::select! {
+                cmd = self.cmds.recv() => {
+                    let cmd = match cmd {
+                        Some(cmd) => cmd,
+                        None => return,
+                    };
+
+                    self.handle_command(sp_addr, cmd, &mut incoming_buf).await;
+                    discovery_idle.reset();
+                }
+
+                result = recv(&self.socket, &mut incoming_buf, &self.log) => {
+                    self.handle_incoming_message(result);
+                    discovery_idle.reset();
+                }
+
+                _ = discovery_idle.tick() => {
+                    debug!(
+                        self.log, "attempting SP discovery (idle timeout)";
+                        "discovery_addr" => %self.discovery_addr,
+                    );
+                    match self.discover(&mut incoming_buf).await {
+                        Ok(addr) => {
+                            if sp_addr != addr {
+                                warn!(
+                                    self.log, "discovered new SP";
+                                    "new_addr" => %addr,
+                                    "old_addr" => %sp_addr,
+                                );
+                            }
+                            sp_addr = addr;
+                        }
+                        Err(err) => {
+                            warn!(
+                                self.log, "idle discovery check failed";
+                                "err" => %err,
+                            );
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    async fn discover(
+        &mut self,
+        incoming_buf: &mut [u8; SpMessage::MAX_SIZE],
+    ) -> Result<SocketAddrV6> {
+        let (addr, response) = self
+            .rpc_call(self.discovery_addr, RequestKind::Discover, incoming_buf)
+            .await?;
+
+        let discovery = response.try_into_discover()?;
+
+        let _ = self.sp_addr.send(Some((addr, discovery.sp_port)));
+        Ok(addr)
+    }
+
+    async fn handle_command(
+        &mut self,
+        sp_addr: SocketAddrV6,
+        command: InnerCommand,
+        incoming_buf: &mut [u8; SpMessage::MAX_SIZE],
+    ) {
+        match command {
+            InnerCommand::Rpc(rpc) => {
+                let result =
+                    self.rpc_call(sp_addr, rpc.kind, incoming_buf).await;
+
+                if rpc.response.send(result).is_err() {
+                    warn!(
+                        self.log,
+                        "RPC requester disappeared while waiting for response"
+                    );
+                }
+            }
+            InnerCommand::SerialConsoleAttach(response_tx) => {
+                let resp = if self.serial_console_tx.is_some() {
+                    Err(Error::SerialConsoleAlreadyAttached)
+                } else {
+                    let (tx, rx) = mpsc::channel(32);
```

Comment:
I really want "effectively unbounded" here, because I don't want this task to block if the receiver isn't pulling messages off of this queue (hence the use of `try_send()` in `forward_serial_console()` instead of `.send().await`). 32 was a WAG at "as long as the receiver is making a reasonable effort to pull messages out at a pace that keeps up with an SP sending them over UDP, we have plenty of buffer to not drop any data".

Reply:
Cool, that works for me. Again, maybe just add a comment. It'd probably be clearer to extract that number into a constant and put the comment over the constant.